[SYCL] Remove tests for ext::oneapi::reduction

aelovikov-intel opened the pull request:
The extension is going to be removed, prepare the test-suite for that.
Also, the removal enables some simplifications/refactoring in the test,
do it here.

aelovikov-intel committed:
[WIP][SYCL] Remove ext::oneapi::reduction

aelovikov-intel committed:
AccDim parameter isn't used anymore (specific to ext::oneapi)

Remove *_s0*.cpp tests, moving some of their test cases into *_s1.

aelovikov-intel committed:
Now that _s1_ doesn't make any sense, rename t files

aelovikov-intel committed:
Remove "#if 0"'ed code

aelovikov-intel committed:
Fuse reduction_nd_range_scalar.hpp/reduction_range_scalar.hpp

aelovikov-intel committed:
Merge reduction_range_scalar.hpp into reduction_utils.hpp

aelovikov-intel committed:
Don't use access::mode to deduce property_list, specify it explicitly

aelovikov-intel committed:
Move test2020USM from reduction_utils.hpp to its single use in reduction_range_usm_dw.cpp

aelovikov-intel committed:
Minor improvement

aelovikov-intel commented:
It might be easier to review each commit in isolation, rather than whole PR at once.

v-klochkov commented on SYCL/Reduction/reduction_nd_N_vars.cpp:
```diff
@@ -17,188 +17,170 @@ constexpr access::mode RW = access::mode::read_write;
-            usm::alloc AllocType4, size_t NWorkItems, size_t WGSize) {
+template <typename T, bool IsUsm, usm::alloc AllocType> auto InitOut(queue &Q) {
+  if constexpr (IsUsm)
+    return (T *)malloc(1, Q, AllocType);
```
IMO, it is good to stay on safer side and keep the checks like "Dev.get_info< info::device::usm_device_allocations >()" before using malloc with device alloc type, etc.

Also, it looks like a mistake here. You alloc too few memory - 1 byte only.
If you want to alloc 1 element, then:
```suggestion
    return malloc<T>(1, Q, AllocType);
```

v-klochkov commented on SYCL/Reduction/reduction_reducer_op_eq.cpp:
```diff
@@ -156,29 +155,28 @@ template <typename T, typename BinaryOperation, OperationEqual OpEq,
   int Error = 0;
-  Error += test<T, true, BinaryOperation, OpEq, IsFP>(Q, Identity);
-  Error += test<T, false, BinaryOperation, OpEq, IsFP>(Q, Identity);
+  Error += test<T, BinaryOperation, OpEq, IsFP>(Q, Identity);
```
testBoth() seems to be unused after the changes, right? If Yes, please remove it.

v-klochkov commented on SYCL/Reduction/reduction_usm.cpp:
```diff
@@ -16,13 +16,13 @@
 using namespace sycl;
 
-template <typename T1, typename T2> class KName;
+template <typename T1, typename T2> class USMKName;
```
This code/test was written when naming lambda functions was required.
If you want to make the code even simpler, then this class can be just removed.

aelovikov-intel replied:
I thought about removing all these names, but then I'd still need to have a test for the cases when unnamed lambda are disabled.

As such, not going to implement as part of this PR, maybe in future.

v-klochkov commented on SYCL/Reduction/reduction_usm_dw.cpp:
```diff
@@ -35,7 +31,7 @@ int test(queue &Q, T Identity, T Init, size_t WGSize, size_t NWItems,
   if (AllocType == usm::alloc::device) {
     Q.submit([&](handler &CGH) {
-       CGH.single_task<KName<Name, class InitKernel>>(
+       CGH.single_task<USMKName<Name, class InitKernel>>(
```
```suggestion
       CGH.single_task(
```

aelovikov-intel replied:
Can't do - crashes in the RT inside program_manager.cpp. Will try to follow-up later.

aelovikov-intel committed:
Update currently disabled tests (verified locally)

aelovikov-intel committed:
Address review comments

aelovikov-intel committed:
Merge remote-tracking branch 'origin/intel' into remove-oneapi-red

v-klochkov commented on SYCL/Reduction/reduction_range_1d_dw.cpp:
```diff
@@ -1,26 +1,22 @@
 
 // This test performs basic checks of parallel_for(range<1>, reduction, func)
-// with reductions initialized with 0-dimensional discard_write accessor
+// with reductions initialized with 1-dimensional discard_write accessor
```
I think this comment is not correct for SYCL 2020. Reduction is not initialized with an accessor, it may be initialized with a buffer and have a property telling that it is initialized with identity value.

v-klochkov replied:
Please fix the comment in other tests as well.

aelovikov-intel committed:
Merge remote-tracking branch 'origin/intel' into remove-oneapi-red

aelovikov-intel committed:
Update comments per review comment

v-klochkov commented on SYCL/Reduction/reduction_range_2d_rw.cpp:
```diff
@@ -7,8 +7,8 @@
-// with reductions initialized with 1-dimensional discard_write accessor
-// accessing 1 element buffer.
+// with reductions initialized with a one element buffer and
+// an initialize_to_identity property.
```
I think this comment was supposed to be used for reduction_*dw tests, not *rw tests, right?

aelovikov-intel replied:
Somehow mixed the rw/dw files. Should be fixed now, thanks!

aelovikov-intel committed:
Swap the comments regarding init/non-init between rw/dw files the way they
supposed to be

aelovikov-intel committed:
Address JIT compilation failure on SYCL_DEVICE_FILTER:acc

Those case were disabled before this PR, I've just accidentally tried to
re-enable them, which I shouldn't have done.

aelovikov-intel committed:
Revert unintended change

aelovikov-intel committed:
std::tuple -> sycl::detail::tuple

The former isn't device copyable on Windows...

aelovikov-intel commented:
All the failure are on tests other than those modified (and IIUC all known), so no regressions because of this PR.

@v-klochkov , @steffenlarsen is there anything preventing this from being approved and subsequently merged?

steffenlarsen commented:
Failures unrelated.
